### PR TITLE
LibJS: Async arrow function and async generator parsing

### PR DESCRIPTION
**Pull request by davidot**
I wanted to implement async generator execution in bytecode but I still have to add a bunch of objects and prototypes so I wanted to start with this.

~Draft since I'm still checking some things
Also I need to add tests for async generator but since we can't execute them yet I'll have to do parse only tests for the time being.~ Done.

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionConstructor.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    define_direct_property(vm.names.prototype, global_object.async_generator_function_prototype(), 0);
+
+    // 27.4.2.1 AsyncGeneratorFunction.length, https://tc39.es/ecma262/#sec-asyncgeneratorfunction-length
+    define_direct_property(vm.names.length, Value(1), Attribute::Configurable);
+
+    // 27.4.2.2 AsyncGeneratorFunction.prototype, https://tc39.es/ecma262/#sec-asyncgeneratorfunction-prototype
+    define_direct_property(vm.names.prototype, global_object.async_generator_function_prototype(), 0);
+}
+
+// 27.4.1.1 AsyncGeneratorFunction ( p1, p2, … , pn, body ), https://tc39.es/ecma262/#sec-asyncgeneratorfunction-constructor
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionConstructor.cpp:36]: we usually link to the same thing for call and construct

```suggestion
// 27.4.1.1 AsyncGeneratorFunction ( p1, p2, … , pn, body ), https://tc39.es/ecma262/#sec-asyncgeneratorfunction
```

File: Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionConstructor.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2021, David Tuin <davidot@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibJS/AST.h>
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionConstructor.h:9]: Why is this included?

File: Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionPrototype.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <LibJS/Runtime/Object.h>
+
+namespace JS {
+
+class AsyncGeneratorFunctionPrototype final : public Object {
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionPrototype.h:13]: This should probably inherit from `PrototypeObject`?

**Review by linusg — CHANGES_REQUESTED** [review]
The second commit message needs refining, it confuses "prototype" with "prototype property" - async functions *do* have a prototype, but unlike regular functions they don't have a prototype property.

`LibJS: Parse Async generator functions` -> `LibJS: Parse async generator functions` for the third commit, c o n s i s t e n c y :^)

**Comment by davidot** [comment]
Fixed all the comments, are the commit messages alright? I'm not super happy with the second one but don't want to write 10 things of which 2 are really relevant to the actual diff.

File: Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionPrototype.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+#include <LibJS/Runtime/AsyncGeneratorFunctionConstructor.h>
+#include <LibJS/Runtime/AsyncGeneratorFunctionPrototype.h>
+#include <LibJS/Runtime/GlobalObject.h>
+
+namespace JS {
+
+AsyncGeneratorFunctionPrototype::AsyncGeneratorFunctionPrototype(GlobalObject& global_object)
+    : Object(*global_object.function_prototype())
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionPrototype.cpp:14]:
```suggestion
    : PrototypeObject(*global_object.function_prototype())
```

**Comment by linusg** [comment]
Build failure looks like a missing include.

> are the commit messages alright?

Looks good, thanks :+1:

File: Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionPrototype.cpp (hunk 4 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Runtime/AsyncGeneratorFunctionPrototype.cpp]: Yeah pushed a little too early

**Review by linusg — APPROVED** [review]

**Pull request merged**